src/EditorFeatures/Core/Implementation/CommentSelection/ToggleBlockCommentCommandHandler.cs

```diff
@@ -0,0 +1,456 @@
+            /// </summary>
+            private IEnumerable<TextSpan> GetUncommentedSpansInSelection()
+            {
+                var uncommentedSpans = new List<TextSpan>();
+
+                // Invert the commented spans to get the uncommented spans.
+                var spanStart = SelectedSpan.Start;
+                foreach (var commentedSpan in IntersectingBlockComments)
+                {
+                    if (commentedSpan.Start > spanStart)
+                    {
+                        // Get span up until the comment and check to make sure it is not whitespace.
+                        var possibleUncommentedSpan = TextSpan.FromBounds(spanStart, commentedSpan.Start);
+                        if (!IsSpanWhitespace(possibleUncommentedSpan))
+                        {
+                            uncommentedSpans.Add(possibleUncommentedSpan);
+                        }
+                    }
+
+                    // The next possible uncommented span starts at the end of this commented span.
+                    spanStart = commentedSpan.End;
+                }
+
+                // If part of the selection is left over, it's not commented.  Add if not whitespace.
+                if (spanStart < SelectedSpan.End)
+                {
+                    var uncommentedSpan = TextSpan.FromBounds(spanStart, SelectedSpan.End);
+                    if (!IsSpanWhitespace(uncommentedSpan))
+                    {
+                        uncommentedSpans.Add(uncommentedSpan);
+                    }
+                }
+
+                return uncommentedSpans;
+            }
+        }
+
+        private class ToggleBlockCommentDocumentDataProvider : IToggleBlockCommentDocumentDataProvider
+        {
+            private readonly ITextSnapshot _snapshot;
+            private readonly CommentSelectionInfo _commentInfo;
+
+            public ToggleBlockCommentDocumentDataProvider(ITextSnapshot textSnapshot, CommentSelectionInfo commentInfo)
+            {
+                _snapshot = textSnapshot;
+                _commentInfo = commentInfo;
+            }
+
+            public int GetEmptyCommentStartLocation(int location)
+            {
+                return location;
+            }
+
+            public IEnumerable<TextSpan> GetBlockCommentsInDocument()
+            {
+                var allText = _snapshot.AsText();
+                var commentedSpans = new List<TextSpan>();
+
+                var openIdx = 0;
+                while ((openIdx = allText.IndexOf(_commentInfo.BlockCommentStartString, openIdx, caseSensitive: true)) > 0)
+                {
+                    // Retrieve the first closing marker located after the open index.
+                    var closeIdx = allText.IndexOf(_commentInfo.BlockCommentEndString, openIdx + _commentInfo.BlockCommentStartString.Length, caseSensitive: true);
+                    // If an open marker is not found (-1), no point in continuing.
```

Comment:
or, we consider it a unterminated comment .  but i'm not going to press on that.